include/deal.II/fe/fe_poly.h

```diff
@@ -278,10 +278,38 @@ class FE_Poly : public FiniteElement<dim,spacedim>
      * then only have to apply the transformation (which is a matrix-vector
      * multiplication) when visiting an actual cell.
      */
+
     std::vector<std::vector<Tensor<1,dim> > > shape_gradients;
+    /**
+     * Array with shape function hessians in quadrature points. There is one
+     * row for each shape function, containing values for each quadrature
+     * point.
+     *
+     * We store the hessians in the quadrature points on the unit cell. We
+     * then only have to apply the transformation when visiting an actual cell.
+     */
+    std::vector<std::vector<Tensor<2,dim> > > shape_hessians;
+
+    /**
+     * Scratch array to store temporary values during hessian calculations in
+     * actual cells.
+     */
+    mutable std::vector<Tensor<2,dim> > untransformed_shape_hessians;
   };
 
   /**
+   * Correct the hessian in the reference cell by subtracting the term corresponding
+   * to the Jacobian gradient for one degree of freedom.
+   */
+  virtual
+  void
+  correct_untransformed_hessians (VectorSlice< std::vector<Tensor<2, dim> > >                       uncorrected_shape_hessians,
```

Comment:
Is there a reason to make this function virtual? Wouldn't it only be used in the internal workings of this class?